docarray/utility/find.py

```diff
@@ -0,0 +1,321 @@
+        from docarray.utility.find import find
+
+
+        class MyDocument(Document):
+            embedding: TorchTensor
+
+
+        index = DocumentArray[MyDocument](
+            [MyDocument(embedding=torch.rand(128)) for _ in range(100)]
+        )
+
+        # use DocumentArray as query
+        query = DocumentArray[MyDocument](
+            [MyDocument(embedding=torch.rand(128)) for _ in range(3)]
+        )
+        results = find(
+            index=index,
+            query=query,
+            embedding_field='tensor',
+            metric='cosine_sim',
+        )
+        top_matches, scores = results[0]
+
+        # use tensor as query
+        query = torch.rand(3, 128)
+        results, scores = find(
+            index=index,
+            query=query,
+            embedding_field='tensor',
+            metric='cosine_sim',
+        )
+        top_matches, scores = results[0]
+
+    :param index: the index of Documents to search in
+    :param query: the query to search for
+    :param embedding_field: the tensor-like field in the index to use
+        for the similarity computation
+    :param metric: the distance metric to use for the similarity computation.
+        Can be one of the following strings:
+        'cosine_sim' for cosine similarity, 'euclidean_dist' for euclidean distance,
+        'sqeuclidean_dist' for squared euclidean distance
+    :param limit: return the top `limit` results
+    :param device: the computational device to use,
+        can be either `cpu` or a `cuda` device.
+    :param descending: sort the results in descending order.
+        Per default, this is chosen based on the `metric` argument.
+    :return: a list of named tuples of the form (DocumentArray, Tensor),
+        where the first element contains the closes matches for each query,
+        and the second element contains the corresponding scores.
+    """
+    if descending is None:
+        descending = metric.endswith('_sim')  # similarity metrics are descending
+
+    embedding_type = _da_attr_type(index, embedding_field)
+
+    # get framework-specific distance and top_k function
+    metric_fn = _get_metric_fn(embedding_type, metric)
+    top_k_fn = _get_topk_fn(embedding_type)
+
+    # extract embeddings from query and index
+    index_embeddings = _extraxt_embeddings(index, embedding_field, embedding_type)
+    query_embeddings = _extraxt_embeddings(query, embedding_field, embedding_type)
+
+    # compute distances and return top results
+    dists = metric_fn(query_embeddings, index_embeddings, device=device)
+    top_scores, top_indices = top_k_fn(
+        dists, k=limit, device=device, descending=descending
+    )
+
+    results = []
+    for indices_per_query, scores_per_query in zip(top_indices, top_scores):
+        docs_per_query = []
+        for idx in indices_per_query:  # workaround until #930 is fixed
+            docs_per_query.append(index[idx])
+        docs_per_query = DocumentArray(docs_per_query)
```

Comment:
why not just add a this in a DocumentArray directly since DocumentArray is just a list of document

Reply:
don't we need to do DocumentArray[smth] here ?

Reply:
>why not just add a this in a DocumentArray directly since DocumentArray is just a list of document
Because `DocumentArray()` doesn't work. Maybe `DocumentArray.empty(0)`?
>don't we need to do DocumentArray[smth] here ?
Yes, good catch